libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp

```diff
@@ -0,0 +1,331 @@
+                    consumed_mahs.erase(consumed_mahs.begin());
+                }
+
+                _state.consumed_mah = avg(consumed_mahs);
+                // _state.consumed_mah = consumed;
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+
+// read_block - returns number of characters read if successful, zero if unsuccessful
+uint8_t AP_BattMonitor_SMBus_SUI::read_block(uint8_t reg, uint8_t* data, uint8_t max_len, bool append_zero) const
+{
+    uint8_t buff[max_len+2];    // buffer to hold results (2 extra byte returned holding length and PEC)
+
+    // read bytes
+    if (!_dev->read_registers(reg, buff, sizeof(buff))) {
+        return 0;
+    }
+
+    // get length
+    uint8_t bufflen = buff[0];
+
+    // sanity check length returned by smbus
+    if (bufflen == 0 || bufflen > max_len) {
+        return 0;
+    }
+
+    // check PEC
+    uint8_t pec = get_PEC(AP_BATTMONITOR_SMBUS_I2C_ADDR, reg, true, buff, bufflen+1);
+    if (pec != buff[bufflen+1]) {
+        return 0;
+    }
+
+    // copy data (excluding PEC)
+    memcpy(data, &buff[1], bufflen);
+
+    // optionally add zero to end
+    if (append_zero) {
+        data[bufflen] = '\0';
+    }
+
+    // return success
+    return bufflen;
+}
+
+// read_bare_block - returns number of characters read if successful, zero if unsuccessful
+uint8_t AP_BattMonitor_SMBus_SUI::read_block_bare(uint8_t reg, uint8_t* data, uint8_t max_len, bool append_zero) const
+{
+    // read bytes
+    if (!_dev->read_registers(reg, data, max_len)) {
+        return 0;
+    }
+
+    // return success
+    return max_len;
+}
+
+AP_BattMonitor_WithCellCount::AP_BattMonitor_WithCellCount(AP_BattMonitor &mon,
+                                                   AP_BattMonitor::BattMonitor_State &mon_state,
+                                                   AP_BattMonitor_Params &params,
+                                                   AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev,
+                                                   uint8_t num_cells)
+    : AP_BattMonitor_SMBus_SUI(mon, mon_state, params, std::move(dev), num_cells)
+{}
+
+void AP_BattMonitor_WithCellCount::timer() {
+    AP_BattMonitor_SMBus_SUI::timer();
+    read_remaining_capacity();
+}
+
+void AP_BattMonitor_WithCellCount::read_cell_voltages() {
+    uint32_t tnow = AP_HAL::micros();
+
+    // read cell voltages
+    if(_cell_count > 0) {
+        // uint8_t voltbuff[_cell_count * 2];
+        uint8_t voltbuff[8];
+
+        if (read_block(REG_CELL_VOLTAGE, voltbuff, 8, false)) {
+            float pack_voltage_mv = 0.0f;
+
+            for (uint8_t i = 0; i < MIN(4, _cell_count); i++) {
+                uint16_t cell = voltbuff[(i * 2) + 1] << 8 | voltbuff[i * 2];
+                _state.cell_voltages.cells[i] = cell;
+                pack_voltage_mv += (float)cell;
+            }
+
+            if(_cell_count >= 4) {
+                float avg = (pack_voltage_mv / 4);
```

Comment:
Since we have to force a pack read anyways, just take the (actual voltage - accumulated cell voltage) / 2 to get a better and more accurate value.

Reply:
yep, done